include/swift/AST/DiagnosticsSema.def

```diff
@@ -4698,6 +4698,18 @@ ERROR(return_init_non_nil,none,
       "'nil' is the only return value permitted in an initializer",
       ())
 
+ERROR(implicit_throws_super_init_nonthrowing,none,
+      "missing call to superclass's initializer; "
+      "'super.init' requires an explicit call because it's a throwing "
+      "initializer while the subclass's initializer is nonthrowing",
+      ())
+
+ERROR(implicit_throws_super_init_rethrowing,none,
+      "missing call to superclass's initializer; "
+      "'super.init' requires an explicit call because it's a throwing initializer "
+      "while the subclass's initializer can only rethrow errors from its parameters",
+      ())
+
 ERROR(implicit_async_super_init,none,
       "missing call to superclass's initializer; "
       "'super.init' is 'async' and requires an explicit call",
```

Comment:
We want diagnostics that belong to the same family to be consistent and recognizable. In this case, we want the `throws` error to ring a bell with a programmer that once already encountered the `async` error, so I believe the wording in all three of these must be aligned. Should we agree on the new wording, I suggest basing the current changes on a commit that updates the `async` error.

Reply:
I completely agree with your reasoning about keeping them consistent and recognizable. With that in mind I believe we should revert the words I had moved back to their original position and arrive even closer to your suggestion. Something I would consider, though, is replacing "subclass's initializer" with "this initializer". What do you think about doing this? To me, it makes it easier to understand what is being referred on the sentence. If we agree on that we will have:

**For reference, the `async` case**
`missing call to superclass's initializer; 'super.init' is 'async' and requires an explicit call`

**On the nonthrowing case**
`missing call to superclass's initializer; 'super.init' is a throwing initializer and requires an explicit call since this initializer is nonthrowing`

**On the `rethrows` case**
`missing call to superclass's initializer; 'super.init' is a throwing initializer and requires an explicit call since this initializer can only rethrow errors from its parameters`

Reply:
On `rethrows errors from its parameters`: Although I perfectly understand the point of this interpretation colloquially, I have doubts about whether it is technically accurate enough for a diagnostic, and und  to put it that way. I’m more inclined to describe rethrowing as either "can only rethrow errors thrown by a call t
* `'super.init' is a throwing initializer and requires either an explicit call or that this initializer be throwing too`

Reply:
Merging them can be good as it also forces us to write a more succinct message. To me, the ending could be a bit more formal, though. Like the following:
`'super.init' is a throwing initializer and requires either an explicit call or that this initializer is also marked as throwing`

Note that the terminology we're using now could be a bit confusing to beginners if they mark the initializer with `rethrows` because they might think "If it's marked as rethrowing, it's obviously marked as throwing too!". To try to mitigate this interpretation I used the word "also" on "[...] or that this initializer is **also** marked as throwing". Other than that I think this is something we might have to deal with, and the documentation should be there to clarify.

What do you think?

---

No worries! I didn't feel discouraged by your comment at all. And I'm still not completely against my original idea of moving the "requires an explicit call" earlier in the sentence. Your comment just made me wonder if I had a good enough reason to propose a change to the async diagnostic and to expand the scope of discussion of this PR. My conclusion was that I didn't, so I preferred making the new diagnostics conform to the writing of the existing one.

Reply:
> Other than that I think this is something we might have to deal with, and the documentation should be there to clarify.
>
> What do you think?

Agreed. The book defines a throwing function specifically as a function with the `throws` specifier, so I reckon a "go read the book" overtone is well grounded. We may as well go a step further and say `also marked as 'throws'`.

___
> No worries! I didn't feel discouraged by your comment at all. And I'm still not completely against my original idea of moving the "requires an explicit call" earlier in the sentence.

Good to know we’re on the same page. And you’re welcome to revisit this in a future PR!

Reply:
This was actually the version I almost suggested instead of my previous one! Hahah Done.

> And you’re welcome to revisit this in a future PR!
Awesome!